server/pbench/bin/index-pbench

```diff
@@ -321,6 +383,166 @@ class ResultData(object):
+        # called "mean", rather than "value", but apart from that,
+        # iteration data and sample data are very similar.
+
+        if 'latency' in results:
+            # Latency results include a latency['usec'] list (or
+            # latency['clat'|'slat'|???] in the fio case) *and* a
+            # throughput['trans_sec'] list (or throughput['iops_sec']
+            # in the fio case). Each list has elements for "all" and
+            # for each client/server pair (or client?) used in the test.
+            if benchmark == "uperf":
+                latency_sublist = ['usec']
+            elif benchmark == "fio":
+                latency_sublist = ['clat', 'lat', 'slat']
+            else:
+                raise BadResultData
+
+            for sublist in latency_sublist:
+                if src_type != 'sample':
+                    return
+                latency_list = results['latency'][sublist]
+                for idx, latency_el in enumerate(latency_list):
+                    datum_wrapper = ResultData.make_datum_wrapper(latency_el)
+                    tseries = latency_el['timeseries']
+                    for res in tseries:
+                        datum = datum_wrapper
+                        datum.update({"value": res['value']})
+                        src_id = _make_source_id(datum)
+                        yield res['date'], datum, src_id
+
+        if 'throughput' in results:
+            # Throughput results include a trans_sec, Gb_sec or
+            # iops_sec list only, depending on the benchmark (uperf or
+            # fio) and the kind of iteration: latency measurements
+            # include a trans_sec throughput sublist, but throughput
+            # measurements include a GB_sec sublist (for uperf) or
+            # an iops_sec sublist (for fio).
+
+            if benchmark == "uperf":
+                if 'trans_sec' in results['throughput']:
+                    throughput_sublist = ['trans_sec']
+                elif 'Gb_sec' in results['throughput']:
+                    throughput_sublist = ['Gb_sec']
+                else:
+                    raise BadResultData
+            elif benchmark == "fio":
+                throughput_sublist = ['iops_sec']
+            else:
+                raise BadResultData
+
+            for sublist in throughput_sublist:
+                if src_type != 'sample':
+                    return
+                throughput_list = results['throughput'][sublist]
+                for idx, xput_el in enumerate(throughput_list):
+                    datum_wrapper = ResultData.make_datum_wrapper(xput_el)
+                    tseries = xput_el['timeseries']
+                    for res in tseries:
+                        datum = datum_wrapper
+                        datum.update({"value": res['value']})
+                        src_id = _make_source_id(datum)
+                        yield res['date'], datum, src_id
+
+        return
+
+    # XXX - copied from ToolData - DRY?
```

Comment:
+1 to following DRY principle.

Reply:
Did as you suggested to eliminate the duplication with `gen_run_timestamps()`. Also, factored out the `convert_timestamp()` method into an abstract object that both `ResultData` and `ToolData` inherit from, to eliminate *that* duplication.